shared/utility/math/stats/multivariate.hpp

```diff
@@ -0,0 +1,119 @@
+            : mean(mean_), covariance(covariance_), gen(rd()) {
+
+            // Calculate the spectral decomposition of the covariance matrix
+            // This is needed MultivariateNormal::sample
+            Eigen::SelfAdjointEigenSolver<Eigen::Matrix<Scalar, N, N>> solver(covariance_);
+            transform = solver.eigenvectors() * solver.eigenvalues().cwiseSqrt().asDiagonal();
+        }
+
+        /**
+         * @brief Samples a N dimensional random vector from this N dimensional multivariate normal distribution
+         *
+         * @details To sample a random vector, x, from a N dimensional multivariate normal distribution, the following
+         * steps are taken
+         * 1. Find any real matrix A such that AA' = \Sigma
+         *  - If \Sigma is positive-definite then the Cholesky decomposition can be used
+         *  - An alternative is to use the A = U\Lambda^{1/2} obtained from the spectral decomposition \Sigma = U\Lambda
+         *  U^{-1}
+         * 2. Let z be an N dimensional vector whose components are sampled from a standard normal distribution
+         * 3. Let x = \mu + Az
+         * https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Drawing_values_from_the_distribution
+         */
+        [[nodiscard]] Eigen::Matrix<Scalar, N, 1> sample() {
+            return mean + transform * Eigen::Matrix<Scalar, N, 1>().NullaryExpr([this]() { return dist(gen); });
+        }
+
+        /**
+         * @brief Calculate the density of the multivariate normal distribution at the point x
+         *
+         * @details This calculation assumes that the covariance matrix is symmetric and positive definite. In this
+         * case, the multivariate normal distribution is "non-degenerate".
+         *
+         * Details of the formula we are using can be found here
+         * https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Non-degenerate_case
+         *
+         * This function can be used to determine the relative likelihood that x was sampled from this distribution.
+         */
+        [[nodiscard]] Scalar density(const Eigen::Matrix<Scalar, N, 1>& x) {
+            const auto z = Scalar(-0.5) * ((x - mean).transpose() * covariance.inverse() * (x - mean))(0);
+            return std::exp(z) / std::sqrt(std::pow(Scalar(2.0 * M_PI), N) * std::abs(covariance.determinant()));
+        }
+
+    private:
+        /// @brief The mean (\mu) of the multivariate normal distribution
+        Eigen::Matrix<Scalar, N, 1> mean = Eigen::Matrix<Scalar, N, 1>::Zero();
+        /// @brief The covariance (\Sigma) of the multivariate normal distribution
+        Eigen::Matrix<Scalar, N, N> covariance = Eigen::Matrix<Scalar, N, N>::Zero();
+        /// @brief A matrix A such that AA' = \Sigma
+        Eigen::Matrix<Scalar, N, N> transform = Eigen::Matrix<Scalar, N, N>::Zero();
+
+        /// @brief Standard normal distribution. Used in MultivariateNormal::sample
+        std::random_device rd{};
```

Comment:
```suggestion
```

Reply:
What's the thinking behind this suggestion